### PR TITLE
Place asset

### PR DESCRIPTION
**Pull request by kianzarrin**
part of #959 
~~blocked by #971~~

### Introduction: 
NetworkIDs are different when saving an intersection in the asset editor and when placing the intersection in-game. The main challenge in this PR tries to map old NetworkIDs to new NetworkIDs. `PlaceIntersectionUtil` was created to encapsulate this mapping operation.

EDIT: I changed my code so I edited the description below:

### Description
- `BuildingDecration.SavePaths()` saves all existing segments to `BuildingInfo.m_paths[]` one element per segment. We observe this order when saving/loading.
- `BuildingDecration.LoadPaths()` creates new segments based on `BuildingInfo.m_paths[]` on segment per element and in the same order. the new segment ids are stored in a temporary list.
- transpiler retrieves the list of new segment ids before it is cleared.
- I create a map of old network ids to new network ids. then I use the map to transfer traffic rules to the newly created network.
- I moved the code from `RecordIUtil` to `SerilaizationUtil` to be more generic. The code is copy-pasted and the only difference is that SerializationUtil does not use the `IRecordable` cast.

PS: This is the final review for this feature branch and should be functional. Vehicle restrictions are handled by #966

### CODE REVIEW

**Comment by Monniasza** [comment]
Is there any progress?

**Comment by kianzarrin** [comment]
I am just waiting for review. @Monniasza you seem to be really interested in adding TMPE rules to assets. Do you have any workshop items you seek to improve?
The Timed traffic light will probably be 11.7 milestone if kvakvs does his part by 11.6.

**Comment by Monniasza** [comment]
> I am just waiting for review. @Monniasza you seem to be really interested in adding TMPE rules to assets. Do you have any workshop items you seek to improve?
> The Timed traffic light will probably be 11.7 milestone if kvakvs does his part by 11.6.

I don't have any, but diverging diamond, diamond and SPUPCLO interchanges can benefit.

**Comment by kianzarrin** [comment]
Its probably best just to build this branch if you can't wait. It can take
a while.

**Comment by kianzarrin** [comment]
Vehicle restriction support is handled in PR #966. If you want to test vehicle restriction then merge with that branch.
If you are wondering why its being merged with master rather than feature branch is because Vehicle records are not only used when placing asset but also in MoveIT copy pasting and The TMPE clear road selection button in the road info panel.

**Comment by kianzarrin** [comment]
@krzychu124 you can verify lane mapping works by checking if lane arrows are transferred properly.

The speed limit data transfer does not work for MoveIT integration either and fixing it is out of the scope of this review. I know what the problem is (its a typo: `LaneIndex` instead of `laneIndex`)and submit a new PR for it.

**Comment by kianzarrin** [comment]
The code should be fully reviewed before this PR can merge.

**Comment by Monniasza** [comment]
@krzychu124 @kvakvs @DaEgi01 @aubergine10 Please review this PR.

File: TLM/TLM/Patch/_InfoManager/SetModePatch.cs [hunk 1]
```diff
@@ -15,7 +15,10 @@ public static class SetModePatch
         [UsedImplicitly]
         public static void Prefix(InfoMode mode, SubInfoMode subMode)
         {
-            if (RoadSelectionPanels.Root?.RoadWorldInfoPanelExt != null) {
+            if (!RoadSelectionPanels.Root)
```
Comment (kianzarrin) [inline TLM/TLM/Patch/_InfoManager/SetModePatch.cs:18]: Road selection panel does not exist in map editor

Comment (krzychu124) [reply TLM/TLM/Patch/_InfoManager/SetModePatch.cs:18]: Yes, but why you are passing `true/false` in line 39? 
Shouldn't it be `() => RoadSelectionPanels.RoadWorldInfoPanel.Hide` or something else (that's where exception is raised)?

Comment (kianzarrin) [reply TLM/TLM/Patch/_InfoManager/SetModePatch.cs:18]: Setmodepatch is changes  UI for integration with road selection panels and does nothing in msp editor no point executing code if it's useless

**Comment by kianzarrin** [comment]
> As far as I remember TTL(record/restore) is not available, right?

yes #967 . I am waiting for kvavks to boost TTL.

> spam of NetService.PublishSegmentChanges(<ID>) called

If you think this is bad it used to print stack trace as well every time!

**Comment by krzychu124** [comment]
> If you think this is bad it used to print stack trace as well every time!

Entry in the log for each call is not the problem for me. but amount of updates. 
Think about it this way: simple intersection - 4 x 6-lane segment is 24 lanes in total. I really doubt it's necessary to make 1500 updates to make it work 😉

**Pull request merged**